imgpkg-003 Rename images when copying bundles

joaopapereira opened the pull request:
A proposal based on the following issues:
https://github.com/vmware-tanzu/carvel-imgpkg/issues/60
https://github.com/vmware-tanzu/carvel-kbld/issues/79

joaopapereira committed:
Initial commit for this proposal

joaopapereira committed:
Add nested bundles caveats and scenario

Add title for some Use Cases with this proposal
Add description of the Overrides section of the CopyWithRename structure
Reply to some open questions
Add one more reason for not having this implemented from the beginning

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,410 @@
+- Copying OCI Images to different repositories can require different user or authentication. When copying an OCI Image
+  to a particular Repository the user needs to have credentials to do so, in a scenario where the
+  repository `my.repo.io/controller` was previously created by a different user, the current user might not have
+  permission to copy the new OCI Image to that particular Repository.
```
i believe ECR requires users to precreate repositories via ECR API vs dockerhub automatically creating them on first use.

joaopapereira replied:
Yes this is true https://docs.aws.amazon.com/AmazonECR/latest/userguide/docker-push-ecr-image.html

joaopapereira replied:
Going to add this

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,410 @@
+provided in the [Problem Statement section](#Problem-Statement).
+
+With this proposal, we will introduce a mechanism to enable the users of `imgpkg` to decide the Repository where each
+OCI Image will be copied to. This will be archived by providing the `copy` command with a Strategy that `imgpkg` will
```
archived -> achieved

joaopapereira replied:
done

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
````diff
@@ -0,0 +1,410 @@
+```yaml
+apiVersion: imgpkg.carvel.dev/v1alpha1
+kind: CopyWithRename
+copyStrategy: same-repository
````
even though imgpkg isnt k8s specific, we are typically following k8s conventions for configuration. in k8s, enums are typically camelized: e.g. SameRepository.

joaopapereira replied:
fixed

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,410 @@
+
+|Original Registry and Repository|Final Registry and Repository|
+|---|---|
+|origin.io/my-image@sha256:aaaa|destination.io/my-image@sha256:aaaa|
```
can we reformat this so it's not a table. hard to read in the github/markdown.

```
origin.io/my-image@sha256:aaaa
  => destination.io/my-image@sha256:aaaa

origin.io/some/path/my-image@sha256:aaaa
 => destination.io/my-image@sha256:aaaa

origin.io/my-image@sha256:aaaa
 => destination.io/destination/path/my-image@sha256:aaaa

origin.io/user/initialmy-image@sha256:aaaa
 => destination.io/destination/path/my-image@sha256:aaaa
```

im not quite sure how some of these get a `destination/path` and some dont.

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,410 @@
+- `location` Is the location where this OCI Image was copied to.
+
+When copying a Bundle between Registries and/or Repositories this new OCI Image will be created in the destination
+Repository and will be tagged with the tag `imgpkg-images-locator-{Bundle SHA}`. This is not a perfect solution since
```
let's format tag as `sha256-abc....imgpkg.locations` (follows similar scheme as https://github.com/sigstore/cosign/issues/8).

joaopapereira replied:
done

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,410 @@
+can see by this example it assumes that in the `.imgpkg/images.yml` file there is 1 entry that will
+match `public-reg.io/simple-app@sha256:*`
+
+#### Nested bundles considerations
```
are we saying that CopyWithRename's strategy/overrides would apply to all nested bundles as well?

joaopapereira replied:
Yes I would make that more explicit

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
````diff
@@ -0,0 +1,410 @@
+images:
+  - origin: world.io/img3@sha256:aaaaaaaaaa
+    location: yet.another.registry.io/nested-bundle-2@sha256:aaaaaaaaaa
+```
````
minor tweaks:

```yaml
apiVersion: imgpkg.carvel.dev/v1alpha1
kind: ImageLocations
images:
- image: world.io/img3@sha256:aaaaaaaaaa
  locations:
  - yet.another.registry.io/nested-bundle-2@sha256:aaaaaaaaaa
```

joaopapereira replied:
done

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,410 @@
+apiVersion: imgpkg.carvel.dev/v1alpha1
+kind: CopyWithRename
+copyStrategy: maintain-repository
+overrides:
```
does the first override win?

joaopapereira replied:
Do you mean if there are 2 overrides that match the same OCI Image?

cppforlife replied:
yup

joaopapereira replied:
The first match should win. Going to add that information

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,410 @@
+kind: CopyWithRename
+copyStrategy: same-repository
+overrides:
+  - source:
```
i think it might be good to have this be consistent with https://carvel.dev/kbld/docs/latest/config/#matching-images (kbld might have to change a bit to support more complex matches)

joaopapereira replied:
done

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,410 @@
+  - source:
+      matchRegistryRepo:
+        registry: public-reg.io
+        repository: simple-app
```
is repository an exact match?

joaopapereira replied:
Yes, I have this comment a couple of lines above

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,410 @@
+apiVersion: imgpkg.carvel.dev/v1alpha1
+kind: CopyWithRename
+copyStrategy: same-repository
+overrides:
```
instead of trying to support potentially large number of rules which still may not be flexible enough, should we allow specifying an inline ytt function that would take an input and produce an output destination.

```
mappingFunctionYttStar: |
  def process(location):
    return ...
  end
```

we could provide a few helper methods that make it easier to deal with image locations.

joaopapereira replied:
Added this option to the document

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
````diff
@@ -0,0 +1,410 @@
+
+The following example contains an example of the configuration YAML Document that would be provided to `imgpkg`
+
+```yaml
````
tweak:

```
apiVersion: imgpkg.carvel.dev/v1alpha1
kind: CopyConfig
strategy: SameRepository
...
```

joaopapereira replied:
I made this change but went 1 step further and changed the strategy names as they were a bit confusing

joaopapereira replied:
Done

joaopapereira committed:
Changes based on initial proposal review

Add new matching capability
Add implementation split into phases
Add ECR Repository creation constraint
Change names of the strategies to be easier to understand

braunsonm commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
````diff
@@ -0,0 +1,410 @@
+
+```yaml
+source:
+  matchExact: public-reg.io/exact-app@sha256:aaaaaaa
````
Is it possible to use a non-sha value here? If I'm wanting to preserve `/my/path/hello-world:v1` I might not know the sha value until after imgpkg generates the `images.yml`

joaopapereira replied:
I just added the possibility of `matchExact` to be done via sha or tag

braunsonm commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
````diff
@@ -0,0 +1,410 @@
+destination: my.private-registry.io/myname/simple-app
+```
+
+Result:
````
Please add an example about how this affects tagging. If I copy two tags from the same repository then how will it be named in the destination. Will it only use a sha value tag or can it copy over the original tag name?

braunsonm commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,410 @@
+source:
+  matchRegistryRepo:
+    registry: public-reg.io
+    repository: simple-app
```
Can I match a tag as well?

joaopapereira replied:
This matcher is just a helper to match against both registry and repository. I assume that if you want something more specific the `matchExact` would be the best option since it supports a match with SHA or with Tag.

joaopapereira committed:
Add more information about the Use Cases

joaopapereira committed:
Add annotation to store tags

Add a new Use Case to ensure `imgpkg` can tag images when copies them
Add requirement for `kbld` to add annotations when creating ImagesLock
Add functions to `ytt` starlark library to allow check for tags and
retrieve the registry location for an image

joaopapereira committed:
Add schema information

Add more schema information about types and what fields are required
Answer one question
Update the name of the matchers to keep consistency with `kbld`
Add note about not allowing `overrides` and `mappingFunctionYttStar` to
be present in the same configuration

joaopapereira committed:
Enhance the text on multiple points

- Make more clear some restrictions
- Rewords parts of the proposal
- Correct some notation

joaopapereira committed:
Change proposal status to `in-review`

gcheadle-vmware commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,740 @@
+- **Repositories**: Place inside a Registry where an OCI Image is located
+- **Registry**: Server that will be used to store and retrieve OCI Images (ex: goharbor.io(on prem), gcr.io(cloud),
+  hub.docker.io(cloud))
+- **OCI**: Open Container Initiative, [the official website](https://opencontainers.org/)
```
Small nit: OCI should be first on this list since it is used in the definition of other terms.

joaopapereira replied:
done

joaopapereira commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
````diff
@@ -0,0 +1,740 @@
+```
+
+In the case of exact matching with tag `imgpkg` will have to do the matching using both the Registry+Repository and the
+annotation `imgpkg.carvel.dev/image-tags` to determine if a particular OCI Image match.
````
During the weekend I was thinking about this annotation and think that this might make more sense as parte of the configuration in ImagesLock.
Since this is a property that imgpkg will use actively to determine which tags to add to images and to do some matching for the copy in this proposal.
iMHO annotations field should be used to store information that imgpkg does not care about or need to do their normal work.
What do you think?

cppforlife replied:
one though i had around this is that if we add tags as a first class field, it's almost like we are telling folks that consume imgpkg that you should be able to access image at a tag location. this isnt something we can always guarantee and that could change as images are pushed to registry. to me that's why it feels more like an annotation. we also always have an option to "upgrade" annotation to a field.

joaopapereira replied:
Lets ignore this for now and we will see if this can be done in the future

joaopapereira committed:
Update order of terminology

DennisDenuto commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,740 @@
+## Terminology / Concepts
+
+- **OCI**: Open Container Initiative, [the official website](https://opencontainers.org/)
+- **Repositories**: Place inside a Registry where an OCI Image is located
```
This definition is going to influence probably most of my comments and thoughts.

[This is what I think a repository is](https://stevelaskerblog.files.wordpress.com/2020/02/registry-repo-tag-ref.png)

If I should change what I think a repo is, lemme know!

joaopapereira replied:
Major revamp here. Should look better now

DennisDenuto commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,740 @@
+- `overrides` (optional; array of overrides) This is a list of overrides for the `strategy`. In this list, the user can
+  define specific behavior for a particular set of OCI Images. Check the [Overrides section](#overrides) for more
+  information on overrides.
+- `mappingFunctionYttStar` (optional; `ytt` starlark function) This field contains an implementation of a `ytt` starlark
```
Is there a difference between a ytt starlark function and a starlark function?

Wondering why call out `ytt`? It seems a bit out of place because I think of `yaml` when I see ytt, and I *think* this is used to operate a mapping function on a string value.

cppforlife replied:
> Is there a difference between a ytt starlark function and a starlark function?

yup. starlark within ytt requires `end` closers.

DennisDenuto replied:
ah makes sense then!

DennisDenuto commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,740 @@
+**Example:**
+
+- OCI Image from `origin.io/my-image@sha256:aaaa` will be copied to `destination.io/my-image@sha256:aaaa`
+- OCI Image from `origin.io/some/path/my-image@sha256:aaaa` will be copied to `destination.io/my-image@sha256:aaaa`
```
I was surprised that the 'repo' chosen as the destination was `my-image` and not `some/path/my-image`.

I'm working with the following assumption: 
- images that are located in repo's that have namespaces, have done so intentionally to disambiguate based on attributes *not* related to version or platform/arch. [1]
- it isn't uncommon for repos with namespaces to contain the same artifact name. (i.e. `runtime/fx` and `sdk/fx`) 
- This choice of not including the namespace was driven by the inconsistency between registries handling of namespaces within a repo name. 

Concerns:
- user (like me) surprised by this behavior. Can they find out by running a dry-run or only after they finish copying?
- loss of contextual information contained in the unique repo name that could be useful in debugging
- losing the unique name of the image can result in multiple artifacts being lumped into the same repo. (i.e. `runtime/fx` and `sdk/fx` would both be sent to `fx`). Which is similar to the problem we are trying to solve.

Suggestions:
- Error out if namespaces cannot be kept when copying to destination

1. Read: Namespace & Repo overlap https://stevelasker.blog/2020/02/17/registry-namespace-repo-names/

joaopapereira replied:
I agree that this is not ideal at all and that we lose information with this copy. To the best of my knowledge, there is not an easy way to know if your registry supports Namespacing or not, you will have to push an image. (If there is a way then my next point is mute) Erroring out in the middle of the copy doesn't help because you might have already copied some images, and now part of the bundle is in the new registry and another part is not.

Another issue here would be if we keep the namespace by default, the probability of bumping into an authentication problem becomes much higher.

Maybe the name of the strategy is the problem here,  maybe it should be called `MaintainOriginImageName` instead of `MaintainOriginRepository`. This way it better aligns with the behavior that we are describing. What do you think?

DennisDenuto replied:
1. I think `MaintainOriginImageName ` is a more accurate description of what the Copy strategy is going to perform. However, as you even point out, it does lose useful information about the unique image name (repository) that the image has. I can't think of a good way to test (upfront) whether a registry supports namespacing, or to test whether the credentials are valid for that repo's namespace.

Here are a couple of random thoughts, which we aren't sold on, but would love your thoughts :-)

1)

> Erroring out in the middle of the copy doesn't help because you might have already copied some images, and now part of the bundle is in the new registry and another part is not

- perhaps a rollback strategy is needed if copy fails. i.e. delete images 😨 that were copied up to the point of failure.  

2) 
- Perhaps on first attempt, try and use a repo with a namespace, and it that doesn't work, try a fallback strategy on changing the namespaces to a single repo name. for e.g. if `some/path/my-image` fails, it would try again with `some-path-my-image`.
  - the 'mangled' name also does lose some information. namely its original repo name, since you are unable to determine what the original namespace was.

joaopapereira replied:
> perhaps a rollback strategy is needed if copy fails. i.e. delete images 😨 that were copied up to the point of failure. (The user might not have delete permissions... so fail the rollback)

We should not be deleting images from the registry. Imagine the case where the blob was already in the registry, we cannot simply delete it from there.

> Perhaps on first attempt, try and use a repo with a namespace, and it that doesn't work, try a fallback strategy on changing the namespaces to a single repo name. for e.g. if some/path/my-image fails, it would try again with some-path-my-image.

In terms of this option, I do not have a strong opinion about it. Feel like this will make our implementation more complicated.
1. `imgpkg` would have to provide a list of possible destination URL and try to understand what error the registry is giving us. We would need some exploration to see what this error is
2. `imgpkg` would have to ensure that we save the correct URL for that particular OCI Image in the new Location OCI Image

Not sure if this is worth it.

What do y'll think about this @cppforlife @pivotaljohn @ewrenn8

cppforlife replied:
doing rollbacks of this sort seems to be quite complicated, not sure if we need it.

i think there was a suggestion around dry-run to see what destinations will be used -- that would be pretty nice to make sure you know where things go before actually running it.

ewrenn8 replied:
I don't think docker hub even supports deleting images from anything other than the browser ui, so a rollback option may not even be possible.

joaopapereira replied:
Can we reach a rough consensus that we will follow the initial proposed solution (disregard the namespace of a particular image) but change the name of the strategy to `MaintainOriginImageName `?

pivotaljohn replied:
I'm chewing over a general approach to addressing any failures that results from a rename...

In addition to ensuring there's a dry-run option 👍🏼 

What if renaming is done on a _best effort_ basis? ... if the rename of an image fails, then _for that image_: 
- `imgpkg` uses the default rename strategy (i.e. `SingleRepository`)
- `imgpkg` tags the copied image containing the desired path+repo name?

Doing so seems to:
- ensure that the copy will always succeed (at least for naming reasons), even if some desired renames don't work as desired
  - this seems particularly desirable in automation cases where a lower-fidelity rename is likely preferred to a failed job.
- allows the operator to make use of namespaces when copying to registries that support them
- mitigates the symptoms that initiated this request for feature (in a conversation with Braunson, they indicated that "Your solution is 'ok' but not 'great'. It's acceptable.")

ewrenn8 replied:
Sorta goes along with my comment below, but if the strategy name is `MaintainOriginImageName`, we will be introducing another term for the part of the url that is the repo.

braunsonm replied:
Following up on this after discussions with @joaopapereira on Slack. It would still be nice to have `MaintainOriginRepository` as a compliment to `MaintainOriginImageName` from a usability point of view.

If this becomes `MaintainOriginImageName` you will need to specify an override for every image in your bundle in order to achieve the functionality of `MaintainOriginRepository`. There are still many private registries that allow you to specify the namespace. Just because some public ones don't I don't think should hamper the flexibility of this proposal.

joaopapereira replied:
Ok, so made a major change that changed the names of the strategies to better indicate what each one is trying to do.

I added the following question in the bottom of the document that would capture the fail or not fail question, here is what is in the proposal:

- What should `imgpkg` do if the User tries to use a strategy that doesn't work on a particular Registry. The most common example would be that the Registry does not support Namespaces and the User is asking `imgpkg` to add them.

  **Answer:** `imgpkg` should error out and tell the user that the Strategy chosen does not work in the destination
  Registry. Instead of trying to do the best effort, in this case, the User is asking `imgpkg` to do an operation that is not supported by the destination Registry. We assume that if this is considered an error for the Registry it should also be an error for `imgpkg`

DennisDenuto commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,740 @@
+- OCI Image from `origin.io/some/path/my-image@sha256:aaaa` will be copied to `destination.io/my-image@sha256:aaaa`
+
+When copying a Bundle to `destination.io/some/path/bundle` all OCI Images will be copied
+to `destination.io/some/path/{Original Repository Name}`
```
This is probably tied to what the definition of a repository is. I would have assumed that the destination would be `destination.io/{Original repo name}`. 

To me the namespace `/some/path/` communicates information *only* for the OCI bundle, not the images it references.

joaopapereira replied:
Check my comment in the prior review point

DennisDenuto replied:
Check my reply to your comment in the prior review point

pivotaljohn replied:
👍🏼  I see value in amplifying [Steve Lasker's solid attempt at disambiguating terms](https://stevelasker.blog/2020/02/17/registry-namespace-repo-names/):
- having a crisp consistency will help not just in naming API symbols, but in documentation as well.
- Steve's steeped in the topic and very involved in a serious creative effort in this domain (Notary V2, others?); so I infer some competency to his choices (and they all resonate with me, personally)

I think as Dennis pointed out, above, Steve also motivates a use-case for making sure that namespaces are supported (search for heading, "Namespace & Repo overlap") in the article:

> Potential life cycle of the web:v1.0 artifact. In the following example, the same web:v1.0 artifact is copied across several repos, using different namespaces, as it makes it way through production, to an archived state.
>
> - `org.example.com/dev/blue-team/web:v1.0`
> - `org.example.com/staging/marketing/web:v1.0`
> - `org.example.com/prod/marketing/web:v1.0`
> - `org.example.com/archive/marketing/web:v1.0`

DennisDenuto commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,740 @@
+
+##### repository
+
+Function will remove all information from an OCI Image location and return the Repository name.
```
Ok, assuming we want to keep namespace and repositories as separate things, there might be value in returning the repository *and* the namespace from this function. (Or perhaps a function to get access to the namespace `some/deep/path`)

This would be useful for folks that *do* care about preserving the namespace. They might decide to change the destination 'repo' name to `some-deep-path-my-image`

joaopapereira replied:
Updated the function names

DennisDenuto replied:
Hmmm I don't see the 'updated' function names. When I checked, the function is still named `repository`.

joaopapereira replied:
Strange can you check again. I replaced `repository` with `imageName` and added a `namespace` function

DennisDenuto replied:
ok I see it. thanks! Resolving this convo

DennisDenuto commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,740 @@
+
+### Use Cases
+
+#### As a User When I copy a Bundle I want to ensure tags present in ImagesLock file are preserved
```
I think I missed this, but where / why are tags present in the `ImagesLock` file?

instead of modifying the image to store references to tags, why not copy every tag associated to the bundle, via the registry `list tags` API? (https://github.com/opencontainers/distribution-spec/blob/main/spec.md#content-discovery)

The user might only be interested with a subset of the tags referencing their image. But they will have the tags they want copied!

joaopapereira replied:
This use case is not depicting "ensure that all the tags on a particular image are preserved". What we want to ensure is that Tags provided as part of the ImagesLock file are added to the OCI Image in the Destination Registry. These tags can be present in the Origin Registry or not.

DennisDenuto replied:
Ok that makes sense! thanks!

DennisDenuto commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,740 @@
+**How might `imgpkg` help to accomplish this Use Case:**
+
+- Maintaining annotations provided in ImagesLock
+- Allow the user to customize the copy of OCI Images to match a naming convention that could help the user identify the
```
customized copy sounds like it will result in a change of sha. I wonder how the user can feel confident about the integrity of the contents of their image. 

How can they know that only the 'customized' copy to include annotations has indeed changed, since the sha256 of the oci manifest would have changed too?

joaopapereira replied:
updated the verbiage, let me know if it still doesn't make sense

DennisDenuto replied:
Ok, so this should only result in a change of image name, Meaning that the sha of the image does *not* change. 

Is that above correct? If so, then consider this thread 'resolved' :-)

DennisDenuto commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,740 @@
+  original OCI Image using only the Repository name
+- Ensure that tags present in the ImagesLock folder are maintained when OCI Images are copied
+
+#### As a User When I have a Bundle with a big number of OCI Images I want to be able to create the CopyConfig file easily
```
In order to ease development / iteration over the CopyConfig file, perhaps a 'dry run' flag would be useful.

Otherwise iterating would involve copying lots of images, only to find out it copied to the wrong repo. (Oops!)

joaopapereira replied:
What would you expect the "dry run" to do, and what information should it provide to the user?

DennisDenuto replied:
Information that would be useful:

- Source and Destination image names. (User can see how the CopyStrategy determined the destination repo name)
  - The above won't catch *all* the errors a user can run into (i.e. authentication), but it would be useful when working on a Copy strategy, especially when working with a ytt starlark function override.

joaopapereira replied:
something like:
`imgpkg copy -b registry.io/bundle --to-repo other.reg.io/copied-bundle -c strategy.yml --dry-run`

Output:
```
Strategy: SingleRepository

Images will try to copied as follow
registry.io/bundle@sha256:113123 -> other.reg.io/copied-bundle@sha256:113123
registry.io/img1@sha256:5555555 -> other.reg.io/copied-bundle@sha256:5555555
index.docker.io/img2@sha256:666666 -> other.reg.io/copied-bundle@sha256:666666
index.docker.io/img3@sha256:777777 -> other.reg.io/overrided-strategy@sha256:777777
```

DennisDenuto replied:
Love it!

joaopapereira replied:
Do you think it makes sense for us to in some way mark the images that are overriding the strategy, something like
`index.docker.io/img3@sha256:777777 -> other.reg.io/overrided-strategy@sha256:777777 (strategy override)`

DennisDenuto replied:
Personally, I wouldn't be too concerned with that level of information. I just want to make sure the destination is correct. Whether that was calculated by the default strategy, override section or hamsters it doesn't really matter. 

The one case it might help with debugging would be if the `strategy override` section isn't being applied for some reason, and the default strategy is instead taking place. (I can't think of how this might happen), but also, the real problem is that the destination repo would be incorrect, which *would* be made obvious with the output not containing `strategy override`

Also, I have a suspicion that not including `strategy override` in the output, will also slightly simplify the implementation.

joaopapereira committed:
Refine the repository ytt starlark function

Add a definition of `Namespace`
Rename function `repository` to `imageName`
Create function definition `namespace`

joaopapereira committed:
Updated language on Use Cases

Better explain what can be customized when executing the copy

joaopapereira committed:
Update location tag name

joaopapereira commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,740 @@
+- `images[].location` (required; string) Is the location where this OCI Image was copied to.
+
+When copying a Bundle between Registries and/or Repositories this new OCI Image will be created in the destination
+Repository and will be tagged with the tag `sha256-{Bundle SHA}-imgpkg.locations`. This is not a perfect solution since
```
updated

joaopapereira commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,740 @@
+
+- New command that generates a configuration file from a Bundle
+
+  `imgpkg tools generate copy-config -b repo.io/bundle@sha256:aaaaaa -o /tmp/copy-config.yml`
```
This command would read all the images from the bundle and would create our `CopyConfig` kind file with all the Images matching with exact match + the destination similar to the `SingleRepository` strategy.

joaopapereira commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
````diff
@@ -0,0 +1,740 @@
+Example of the override
+
+```yaml
+imageRepo: public-reg.io/simple-app
+destination: my.private-registry.io/myname/simple-app
````
What in particular? the matcher or the `destination`?

cppforlife replied:
destination mostly.

joaopapereira replied:
I updated the name, so I am going to resolve this point here.
Also did some changes to the CopyConfigs and if there is any comment around any of these 2 changes we should start a new thread

joaopapereira commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,740 @@
+
+- OCI Image from `origin.io/my-image@sha256:aaaa` will be copied to `destination.io/some/path/my-image@sha256:aaaa`
+- OCI Image from `origin.io/image/origin/my-image@sha256:aaaa` will be copied
+  to `destination.io/some/path/my-image@sha256:aaaa`
```
This might be a very big name, and by default, I think we should not mangle it. I believe that the override capability described in this proposal would allow the user to mangle the names if they intend to.
I think maybe in the future we can make `imgpkg` a little bit smarter and when it finds 2 images with the same name, it would mangle the last portion of the Namespace of the Image or something similar to that.

cppforlife replied:
> This might be a very big name

are there certain repo fragment length limits?

cppforlife replied:
> I believe that the override capability described in this proposal would allow the user to mangle the names if they intend to.

i have a feeling that most folks do not care that much about individual naming but rather just want something that could be guessed if it shows up in different tools (like vuln scanner)

> I think maybe in the future we can make imgpkg a little bit smarter and when it finds 2 images with the same name

this becomes non-deterministic which is think is a very dangerous road to get on.

joaopapereira replied:
> are there certain repo fragment length limits?

Trying to read the specs to see if there is a limitation

> i have a feeling that most folks do not care that much about individual naming but rather just want something that could be guessed if it shows up in different tools (like vuln scanner)

Yes I agree with that, would you agree that we can archive this via the Overrides section or `imgpkg` should try to do some sort of mangling by default?

> this becomes non-deterministic which is think is a very dangerous road to get on.

I agree that this is not ideal, that is why I was saying that we can eventually in the future iterate on this if we see that there is a better way

pivotaljohn replied:
> I agree that this is not ideal, that is why I was saying that we can eventually in the future iterate on this if we see that there is a better way

I'm down with the idea of deferring design decisions to the last responsible moment. 👍🏼  What comes with that, for me, is that we've made use of the information _available_ to us. I have an intuition that we can do better than non-deterministic behaviors, even out the gate...

joaopapereira replied:
> are there certain repo fragment length limits?

Cannot find it in the spec but after talking with Steve he says that the Repository (/namespace/sub-namespace/repo) has a maximum length of 256

cppforlife replied:
+1 -- we should be deterministic.

joaopapereira committed:
Correct verbiage in the registry URL

joaopapereira committed:
Add dry-run flag to the copy command

joaopapereira committed:
Apply Formatting

joaopapereira committed:
Disallow copy to different registries

Answer one of the open questions.
Add note to overrides and the advanced case to ensure an error is
displayed if the user tries to use different registries

cppforlife commented:
one late constraint ill throw that i just thought of... we should ensure that registry replication can be used with this feature. ie if bundle that uses images.location gets replicated to a different registry, ideally images.location should work in the context of that new registry (assuming that all relevant dependent images were also replicated).

joaopapereira committed:
Add considerations for registry replication scenario

joaopapereira committed:
Rename overrides `destination` to `destinationRepo`

This will remove the Registry information from the strategy
configuration file

joaopapereira committed:
Add more clarification around the ImagesLocation on nested bundle

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,837 @@
+
+**Note:** `copy` and `pull` will have to be able to implement the above search
+
+#### Registry replication considerations
```
the replication case that we have not addressed as far as i can see is:

user ran imgpkg copy ... --to-repo registry.foo.com/apps/project1/bundle-img -c ... so that images are placed into same "directory": registry.foo.com/apps/project1/

registry.foo.com/apps/project1/bundle-img
registry.foo.com/apps/project1/bundle-img:...imgpkg.locations (generated because images were exploded...)
registry.foo.com/apps/project1/img1 (originally came from docker)
registry.foo.com/apps/project1/img2 (originally came from gcr)

then somebody sets up replication rule that says copy everything between:

registry.foo.com/apps/project1/ to
registry.secret.com/remote-project/

which results in something like

registry.secret.com/remote-project/bundle-img
registry.secret.com/remote-project/bundle-img:...imgpkg.locations (generated because images were exploded...)
registry.secret.com/remote-project/img1 (originally came from docker)
registry.secret.com/remote-project/img2 (originally came from gcr)

how do make sure that imgpkg pull can find img1, and img2 within same "directory".

joaopapereira replied:
We did not address that particular scenario. Maybe what we can do is in our Searching Order for OCI Image we can update it to look like the following:

1. Bundle Registry URL + Location specified in the ImagesLocation configuration
1. `registry` field from `ImagesLocation` + Location specified in the `ImagesLocation` configuration
2. Bundle Registry URL + Bundle Location - Bundle Image Name + Image Name 
1. Bundle Registry URL + Bundle Location + @sha256:{Image SHA}
   **Note:** We need to keep this possible location for retro compatibility.
1. Location present in ImagesLock file

Assumptions:
- Bundle being pulled or copied is in the following location `registry.io/some/place/bundle`
- ImagesLock file contains 1 Image that point to `other.registry.io/img1@sha256:aaaaaaa`
- ImagesLocation looks like the following
```
registry: private.registry.io
images:
- image: other.registry.io/img1@sha256:aaaaaaa
   repository: some/other/place/img1
```
`imgpkg` would search the following places for the OCI Image in the order
1. registry.io/some/other/place/img1@sha256:aaaaaaa
2. private.registry.io/some/other/place/img1@sha256:aaaaaaa
3. registry.io/some/place/img1@sha256:aaaaaaa
4. registry.io/some/place/bundle@sha256:aaaaaaa
5. other.registry.io/img1@sha256:aaaaaaa

---

@cppforlife do you think this addresses your scenario?

joaopapereira replied:
This scenario should not be covered in the new draft of the proposal

ewrenn8 commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,837 @@
+strategy: MaintainOriginRepository
+overrides:
+  - image: public-reg.io/exact-app@sha256:aaaaaaa
+    destinationRepo: myname/exact-app
```
I am a bit confused here based on the above discussions. Will this override result in the image being found at `destination.io/myname/exact-app`? What happens when a user copies with the destination as `destination.io/my-namespace` because of the above behavior? Would the image be located at `destination.io/my-namespace/myname/exact-app` or `destination.io/myname/exact-app`?
 
It feels like we have a partial definition of what constitutes the repo portion of a url, which is resulting in us having three different meanings for the term  based on where it is used:
1. `imageRepo` matcher - should contain full url which has registry, namespace, and repo
2. `destinationRepo` - shouldn't contain registry but is allowed to contain both namespace and repo
3.  Preserved repo when using MaintainOriginRepository - only recognizes last portion of url as repo (drops namespace and registry which are included in above defs)

My understanding is that repo is the last segment of the url. For example, based on the terminology section, I would expect `index.docker.io/_/nginx@sha256:...` to have:
- registry: index.docker.io
- namespace: _
- repo: nginx
- image: sha256...

This understanding may be incorrect (and potentially harder to implement because ggcr doesn't have the idea of namespace iirc) but if it is, I think it should be clarified and made consistent wherever the terms are used or else users will have no way of knowing how we've chosen to define repo in a particular location.

joaopapereira replied:
@ewrenn8 I made some changes to the verbiage of the full proposal and tried to make it consistent. Please check it out and if it looks good, please resolve this issue

ewrenn8 commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,837 @@
+
+In the event that 2 Overrides match a particular OCI Image the first Override in the list takes precedence.
+
+In the following example the `exact-app` OCI Image will be copied to `my.private-registry.io/myname/exact-app`
```
This is extremely minor, but can we describe the reason here. It could be easy to miss the digest was left off and think it will be copied to the first override because of the behavior called out above about multiple overrides matching and not because the second override doesn't actually match.

joaopapereira replied:
made an update on the verbiage, do you think it better explains the selection?

ewrenn8 replied:
Yea, looks good!

joaopapereira committed:
Better explanation of the override matching

joaopapereira commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,837 @@
+`imgpkg` only is aware of `registry.io` (the original location of the OCI Image) and `other.registry.io` (the location
+where the Bundle is being pulled)
+
+If the Registry URL to where the Bundle is copied is saved in the `ImagesLocation` file, `imgpkg`, would be able to know
+about the interim registry, allowing for a best effort find of all the OCI Images.
```
> replication is typically used to get content down to a private registry and in that case it's not desirable to try to get stuff from other locations

I feel like the above comment is enough to convince me to rollback the change I proposed. Going to update the proposal to match it

joaopapereira replied:
Made the change on the proposal

joaopapereira committed:
Remove the registry information from ImagesLocation

- Add Use Case for the replicated registry
- Add question about replicated registries

joaopapereira committed:
Add a new order of search for images

Change the way Destination field works

jorgemoralespou commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,887 @@
+   When this is the case `destinationRepo: /here` the OCI Image will be copied to `registry.io/here`. `imgpkg` assumes
+   that the OCI Image should be copied to the root of the Registry
+
+1. the value does not start with `\`
```
This is very subtle, and will be source of confusion if not documented properly.

joaopapereira replied:
Added note to ensure the documentation reflects this well

jorgemoralespou commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
````diff
@@ -0,0 +1,887 @@
+destinationRepo: myname/exact-app
+```
+
+In the case of exact matching with tag `imgpkg` will have to do the matching using both the Registry+Repository and the
````
This seems to me confusing. Probably would benefit from more detailed description of what happens here.

joaopapereira replied:
Updated this description

jorgemoralespou commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,887 @@
+
+Function will remove all information from an OCI Image location and return the Image name.
+
+Definition: `def imageName(image) -> string`
```
Interesting as above this is described as the Repository (although I find that name confusing and I like better image name).

jorgemoralespou commented:
Is there a way to capture in the ImagesLock file (or in a resulting file) the result of the copy so that it can be tracked? Source image, destination image and strategy used on copy.?

joaopapereira committed:
Better explain some parts of the proposal

- Add commands to show how to use the configuration file
- Better explain the exact mathc by tag

joaopapereira committed:
Ensure coherent language on the proposal

- Add Image that helps with terminology
- Convert strategy name `MaintainOriginRepository` to `RelativeToRegistryKeepNamespace`

joaopapereira committed:
Add 2 new copy strategies

Strategies:
- Copy all OCI Images to the new Registry and all paths are relative to Bundle
  Namespace and keeps Image Namespace
- Copy all OCI Images to the new Registry and all paths are relative to Bundle
  Namespace

joaopapereira commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
````diff
@@ -0,0 +1,887 @@
+```
+
+The above configuration will ensure that the OCI Image `public-reg.io/exact-app@sha256:aaaaaaa` will be copied
+to `my.private-registry.io/myname/exact-app@sha256:aaaaaaa`. For other options of matching check
````
done

joaopapereira committed:
Add behavior for `imgpkg` when dealing with large number of namespaces

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,1035 @@
+`corp.gcr.io/projects/project-a/bank-app:v1` explodes into:
+
+- **Registry** (`corp.gcr.io`): Server that will be used to store and retrieve OCI Images
+    - other examples: goharbor.io, gcr.io, index.docker.io
```
may be we should add `foo.com/blah` as another example of registry to indicate that it's not only hostname.

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,1035 @@
+images:
+  - image: public-reg.io/exact-app@sha256:aaaaaa
+    annotations:
+      - imgpkg.carvel.dev/image-tags: [ "will-not-match" ]
```
nit: not an array.

pivotaljohn replied:
I assume you mean `annotations` is not an array, but a map.

joaopapereira replied:
done

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,1035 @@
+
+To help users to create simpler mapping the following functions will be available:
+
+##### regexMatch
```
we should be able to use ytt's regexp module.

joaopapereira replied:
done

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,1035 @@
+
+**Note:** `overrides` and `mappingFunctionYttStar` cannot be present at the same time in a configuration.
+
+#### Copy Strategies
```
i think we should have a bullet list here quickly describing all strategies.

joaopapereira replied:
done

cppforlife commented on proposals/imgpkg/003-copy-bundles-with-rename/Readme.md:
```diff
@@ -0,0 +1,1035 @@
+
+**Note:** `overrides` and `mappingFunctionYttStar` cannot be present at the same time in a configuration.
+
+#### Copy Strategies
```
i think it would be useful to have `RelativeToBundleDashedFullImageName`:

`blah.com/some/ns/image-name` would become `registry.com/another-ns/some-ns-image-name`

pivotaljohn replied:
It was mentioned that it could be useful to provide "portable" configuration (i.e. be able to use the same config for multiple `imgpkg copy` invocations, each targeting a different registry)...

Would we be able to detect when a copy fails because namespaces aren't supported? If so, we _could_ get closer to that portability goal if `RelativeToBundleKeepNamespace` first attempts with the namespace, as is... and if it fails in a definitive way, could downshift to the dashed image name?

jorgemoralespou replied:
I think that the fact that there's a new strategy `RelativeToBundleDashedFullImageName` almost similar to an existing one `RelativeToBundle` to me denotes that multiple variations of strategies will come, and not just to one, but to many, so eventually there will be a explosion of strategies to cope with this subtleties. I would recommend not adding so many strategies but adding additional configuration to the strategies, e.g:

```
apiVersion: imgpkg.carvel.dev/v1alpha1
kind: CopyConfig
strategy: RelativeToBundle
keep: Namespace|Image|Repositiry
dashed: Namespace|Image|Repository
```
So that with some additional attributes you can get multiple different strategies.

joaopapereira committed:
Multiple improvements

- Correct the ImagesLock yaml
- Enumerate the strategies for better visibility
- Start using ytt modules, instead of create things like regexMatch.
  Since we will have access to these module we can take advantage of all
  of them

braunsonm commented:
Any update on this? Would still be great to see.

joaopapereira committed:
Add the gist with a different implementation of the strategies
